contracts/nft_royalty/lib.rs

```diff
@@ -0,0 +1,437 @@
+        RoyaltyMetadataKeyNotFound(Ticker),
+        /// Trying to decode [`NFTArtistRules`] from [`AssetMetadataValue`] failed.
+        FailedToDecodeMetadataValue(String),
+        /// [`NFTOffer::purchase_ticker`] can't be used as royalty payment.
+        TickerNotAllowedForRoyalty(Ticker),
+        /// No [`PortfolioId`] for the given [`IdentityId`].
+        RoyaltyPortfolioNotFound(IdentityId),
+        /// The contract caller has already set up a custodial portfolio.
+        RoyaltyPortfolioAlreadyExists,
+        /// Contract hasn't been initialized.
+        MissingContractInitialization,
+    }
+
+    impl From<PolymeshError> for Error {
+        fn from(err: PolymeshError) -> Self {
+            Self::PolymeshInk(err)
+        }
+    }
+
+    /// A contract that manages non-fungible token transfers.
+    #[ink(storage)]
+    pub struct NftRoyalty {
+        /// Upgradable Polymesh Ink API.
+        api: PolymeshInk,
+        /// The [`AccountId`] that called the contract's constructor.
+        admin: AccountId,
+        /// Venue owned by the contract.
+        venue_id: Option<VenueId>,
+        /// The identity of the contract.
+        contract_identity: IdentityId,
+        /// The portfolios that will receive the royalty value for each identity.
+        royalty_portfolios: Mapping<IdentityId, PortfolioId>,
+        /// The asset metadata key for each ticker.
+        metadata_keys: Mapping<Ticker, u64>,
+    }
+
+    impl NftRoyalty {
+        /// Inititializes the [`NftRoyalty`] storage.
+        #[ink(constructor)]
+        pub fn new() -> Result<Self> {
+            Ok(Self {
+                api: PolymeshInk::new()?,
+                admin: Self::env().caller(),
+                venue_id: None,
+                contract_identity: PolymeshInk::get_our_did()?,
+                royalty_portfolios: Mapping::default(),
+                metadata_keys: Mapping::default(),
+            })
+        }
+
+        /// Inititializes the [`NftRoyalty`] storage with an address of the contract.
+        #[ink(constructor)]
+        pub fn new_with_hash(hash: Hash) -> Result<Self> {
+            Ok(Self {
+                api: PolymeshInk::new_with_hash(hash),
+                admin: Self::env().caller(),
+                venue_id: None,
+                contract_identity: PolymeshInk::get_our_did()?,
+                royalty_portfolios: Mapping::default(),
+                metadata_keys: Mapping::default(),
+            })
+        }
+
+        /// Initializes the contract creating a venue that will be controlled by the contract.
+        #[ink(message)]
+        pub fn initialize_contract(&mut self) -> Result<VenueId> {
+            self.contract_identity = PolymeshInk::get_our_did()?;
+            let venue_id = self
+                .api
+                .create_venue(VenueDetails(VENUE_DETAILS.to_vec()), VenueType::Other)?;
+            self.venue_id = Some(venue_id);
+            Ok(venue_id)
+        }
```

Comment:
Anyone can call this multiple times and keep creating new venues.  There needs to be a guard against this method being called multiple times.  If the `venue_id` changes then all NFT collections would need to update their list of allowed venues, otherwise all future transfers would fail.